### PR TITLE
[ottool] Add more argument checks to transition command 

### PR DESCRIPTION
**Pull request by msfschaffner**
While the generic transition command already works, it does not check any preconditions.
This adds more argument checking to make sure the requested transition is valid, and a token has been provided if one is required.

### CODE REVIEW

File: sw/host/opentitanlib/src/dif/lc_ctrl.rs [hunk 1]
```diff
@@ -93,6 +386,19 @@ impl From<[u8; 16]> for DifLcCtrlToken {
...
+impl From<[u32; 4]> for DifLcCtrlToken {
+    fn from(words: [u32; 4]) -> Self {
+        let mut bytes = [0u8; 16];
+        for i in 0..4 {
+            bytes[4 * i + 0] = (words[i] & 0xFF) as u8;
+            bytes[4 * i + 1] = ((words[i] >> 8) & 0xFF) as u8;
+            bytes[4 * i + 2] = ((words[i] >> 16) & 0xFF) as u8;
+            bytes[4 * i + 3] = ((words[i] >> 24) & 0xFF) as u8;
+        }
+        DifLcCtrlToken::from(bytes)
```
Comment (cfrantz) [inline sw/host/opentitanlib/src/dif/lc_ctrl.rs]: If you wanted to be fancy here, you could try something like this:
```
        let bytes: [u8; 16] = words
            .iter()
            .map(|v| v.to_le_bytes())
            .flatten()
            .collect::<Vec<u8>>()
            .try_into() 
            .unwrap();
        DifLcCtrlToken::from(bytes)
```

Because the input is fixed size and the output is the same size in bytes, the `try_into` is going to succeed and, thus, we can simply unwrap the result.

File: sw/host/opentitanlib/src/dif/lc_ctrl.rs [hunk 2]
```diff
@@ -83,8 +88,296 @@ impl DifLcCtrlState {
             _ => Ok(DifLcCtrlState::StateInvalid),
         }
     }
+
+    /// Checks whether a transition is valid and whether a non-zero token is
+    /// required.
+    pub fn check_transition(self, target: DifLcCtrlState) -> DifLcCtrlTransCheck {
+        match target {
+            // TestUnlocked states
+            DifLcCtrlState::TestUnlocked0 => match self {
```
Comment (nbdd0121) [inline sw/host/opentitanlib/src/dif/lc_ctrl.rs:97]: These double matches should be matching a tuple:
```rust
match (self, target) {
    (DifLcCtrlState::Raw, DifLcCtrlState::TestUnlocked0) => ...,
    (_, DifLcCtrlState::TestUnlocked0) => ...,
}
```

File: sw/host/opentitanlib/src/dif/lc_ctrl.rs [hunk 3]
```diff
@@ -93,6 +386,19 @@ impl From<[u8; 16]> for DifLcCtrlToken {
...
 }
 
+impl From<[u32; 4]> for DifLcCtrlToken {
+    fn from(words: [u32; 4]) -> Self {
+        let mut bytes = [0u8; 16];
+        for i in 0..4 {
+            bytes[4 * i] = (words[i] & 0xFF) as u8;
+            bytes[4 * i + 1] = ((words[i] >> 8) & 0xFF) as u8;
+            bytes[4 * i + 2] = ((words[i] >> 16) & 0xFF) as u8;
+            bytes[4 * i + 3] = ((words[i] >> 24) & 0xFF) as u8;
```
Comment (nbdd0121) [inline sw/host/opentitanlib/src/dif/lc_ctrl.rs:396]:
```suggestion
            bytes[4 * i..][..4].copy_from_slice(&words[i].to_le_bytes());
```

File: sw/host/opentitanlib/src/dif/lc_ctrl.rs [hunk 4]
```diff
@@ -107,6 +413,15 @@ impl DifLcCtrlToken {
             .for_each(|(word, out)| *out = word);
         out_words
     }
+
+    /// Returns 1 if the token is zero.
```
Comment (nbdd0121) [inline sw/host/opentitanlib/src/dif/lc_ctrl.rs:417]:
```suggestion
    /// Returns true if the token is zero.
```

File: sw/host/opentitanlib/src/dif/lc_ctrl.rs [hunk 5]
```diff
@@ -107,6 +413,15 @@ impl DifLcCtrlToken {
...
+    pub fn is_zero(self) -> bool {
+        let mut res = true;
+        for b in self.0.data {
+            res = res && (b == 0);
+        }
+        res
```
Comment (nbdd0121) [inline sw/host/opentitanlib/src/dif/lc_ctrl.rs:423]:
```suggestion
        self.0.data.iter().all(|&x| x == 0)
```
or even
```suggestion
        self.0.data == [0; 16]
```

File: sw/host/opentitantool/src/command/lc.rs [hunk 6]
```diff
@@ -22,12 +22,31 @@ pub struct LcStateReadResult {
...
+        bail!("Invalid transition from {} to {}.", lc_state, target);
+    }
+    if check.token && token.is_zero() {
+        bail!(
+            "Transition from {} to {} requires a non-zero token.",
+            lc_state,
+            target
+        );
+    }
+    if (!check.token) && (!token.is_zero()) {
```
Comment (nbdd0121) [inline sw/host/opentitantool/src/command/lc.rs:44]: Remove the extra parenthesis.

File: sw/host/opentitantool/src/command/lc.rs [hunk 7]
```diff
@@ -22,12 +22,31 @@ pub struct LcStateReadResult {
...
+    target: DifLcCtrlState,
+    token: DifLcCtrlToken,
+) -> Result<()> {
     let lc_state =
         DifLcCtrlState::from_redundant_encoding(jtag.read_lc_ctrl_reg(&LcCtrlReg::LcState)?)?;
-    if lc_state != DifLcCtrlState::Raw {
-        bail!("Device must be in Raw LC state. It is in: {}", lc_state);
+    let check = lc_state.check_transition(target);
+    if !check.valid {
+        bail!("Invalid transition from {} to {}.", lc_state, target);
```
Comment (nbdd0121) [inline sw/host/opentitantool/src/command/lc.rs:35]: These `if !cond { bail!(...) }` could just be `ensure!(cond, ...)`.

Also you can take advantage of implicit capture by just write `bail!("Invalid transition from {lc_state} to {target}.");`

File: third_party/rust/repos.bzl [hunk 8]
```diff
@@ -95,11 +95,13 @@ def rust_repos(rules_rust = None, safe_ftdi = None, serde_annotate = None):
     http_archive_or_local(
         name = "rules_rust",
         local = rules_rust,
-        sha256 = "1a110ae8bfa34091e2a62e4c0325f8231f934b2956f3a585d4cafba5c81c2676",
-        strip_prefix = "rules_rust-rebase-20231115_01",
-        url = "https://github.com/lowRISC/rules_rust/archive/refs/tags/rebase-20231115_01.tar.gz",
+        sha256 = "e85891dd6bb24e877a9622c09f028b9e11c0f4df69db4735194eb1d8dc7ae917",
+        strip_prefix = "rules_rust-rebase-20230822_02",
+        url = "https://github.com/lowRISC/rules_rust/archive/refs/tags/rebase-20230822_02.tar.gz",
```
Comment (nbdd0121) [inline third_party/rust/repos.bzl:100]: The version bump is there to fix an issue found in OT cert gen, please don't downgrade.

File: sw/host/tests/chip/rv_core_ibex_isa/src/main.rs [hunk 9]
```diff
@@ -46,10 +46,10 @@ fn ibex_isa_smoke_test(opts: &Opts, transport: &TransportWrapper) -> Result<()>
     uart.clear_rx_buffer()?;
 
     // Load SRAM program
-    match opts
-        .sram_program
-        .load_and_execute(&mut *jtag, ExecutionMode::JumpAndWait(Duration::from_secs(5)))?
-    {
+    match opts.sram_program.load_and_execute(
```
Comment (nbdd0121) [inline sw/host/tests/chip/rv_core_ibex_isa/src/main.rs:49]: Unrelated change?

File: sw/host/opentitantool/src/command/lc.rs [hunk 10]
```diff
@@ -282,7 +305,22 @@ impl CommandDispatch for Transition {
...
+        // Reset the chip so that LC_CTRL is in a clean state.
+        let _ = transport.reset_target(Duration::from_millis(50), false);
+        std::thread::sleep(Duration::from_millis(50));
+
+        // Check whether this is a valid transition.
         let token_words = parse_token_str(self.token.as_str())?;
+        check_lc_transition(
+            &mut *jtag,
+            self.target_lc_state,
+            DifLcCtrlToken::from(token_words),
```
Comment (nbdd0121) [inline sw/host/opentitantool/src/command/lc.rs:322]: Can you make `parse_token_str` return `DifLcCtrlToken` instead?

File: third_party/rust/repos.bzl (hunk 8 above)
Comment (msfschaffner) [reply third_party/rust/repos.bzl]: Sorry this is a rebase snafu.
I had to change this temporarily on my machine due to https://github.com/lowRISC/opentitan/issues/20577.
I removed it from this PR.

File: sw/host/opentitantool/src/command/lc.rs (hunk 7 above)
Comment (msfschaffner) [reply sw/host/opentitantool/src/command/lc.rs]: Changed to `ensure!`

File: sw/host/opentitantool/src/command/lc.rs (hunk 6 above)
Comment (msfschaffner) [reply sw/host/opentitantool/src/command/lc.rs]: done

File: sw/host/opentitanlib/src/dif/lc_ctrl.rs (hunk 5 above)
Comment (msfschaffner) [reply sw/host/opentitanlib/src/dif/lc_ctrl.rs]: Went for the latter one, since it seems more intuitive.

File: sw/host/opentitanlib/src/dif/lc_ctrl.rs (hunk 4 above)
Comment (msfschaffner) [reply sw/host/opentitanlib/src/dif/lc_ctrl.rs]: done

File: sw/host/opentitanlib/src/dif/lc_ctrl.rs (hunk 3 above)
Comment (msfschaffner) [reply sw/host/opentitanlib/src/dif/lc_ctrl.rs]: Went with a slightly different variant suggested by Chris above.

File: sw/host/opentitanlib/src/dif/lc_ctrl.rs (hunk 2 above)
Comment (msfschaffner) [reply sw/host/opentitanlib/src/dif/lc_ctrl.rs]: done

File: sw/host/opentitanlib/src/dif/lc_ctrl.rs [hunk 11]
```diff
@@ -38,6 +38,11 @@ with_unknown! {
     }
 }
 
+pub struct DifLcCtrlTransCheck {
+    pub valid: bool, // Transition is valid
```
[the remark this answers is not in the document]
Comment (msfschaffner) [reply sw/host/opentitanlib/src/dif/lc_ctrl.rs]: fixed

File: sw/host/tests/chip/rv_core_ibex_isa/src/main.rs (hunk 9 above)
Comment (msfschaffner) [reply sw/host/tests/chip/rv_core_ibex_isa/src/main.rs]: Yeah this is unrelated and went in together with the `repos.bzl` change above - this is removed now.

File: sw/host/opentitantool/src/command/lc.rs (hunk 10 above)
Comment (msfschaffner) [reply sw/host/opentitantool/src/command/lc.rs]: Yes, done.

**Review by cfrantz — APPROVED** [review]

File: sw/host/opentitanlib/src/test_utils/lc_transition.rs [hunk 12]
```diff
@@ -91,17 +92,16 @@ fn setup_lc_transition(
...
-            &LcCtrlReg::TransitionToken0,
-            &LcCtrlReg::TransitionToken1,
-            &LcCtrlReg::TransitionToken2,
-            &LcCtrlReg::TransitionToken3,
-        ];
-
-        for (reg, value) in iter::zip(token_regs, token_words) {
-            jtag.write_lc_ctrl_reg(reg, value)?;
-        }
+    let token_words = token.expect("Get token").into_register_values();
```
Comment (nbdd0121) [inline sw/host/opentitanlib/src/test_utils/lc_transition.rs:95]: This is causing CI to fail. Why expect here?

Comment (msfschaffner) [reply sw/host/opentitanlib/src/test_utils/lc_transition.rs]: I did this due to a compiler suggestion.

But I ended up reverting the `trigger_lc_transition` interface change in `lc_transition` since it affects many other existing provisioning programs and I would like to avoid changing these at this point.
